BUG: Enforce correct encoding in stata

bashtage opened the pull request:
Ensure StataReader and StataWriter have the correct encoding.
Standardized default encoding to 'latin-1'

closes #15723

 - [x] closes #15723
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

bashtage committed:
BUG: Enforce correct encoding in stata

Ensure StataReader and StataWriter have the correct encoding.
Standardized default encoding to 'latin-1'

closes #15723

jreback commented:
@bashtage side issue. We finally have 32-bit daily wheels.

https://travis-ci.org/MacPython/pandas-wheels/jobs/213101390

so this one came up in testing (I am working fixing the rest). Unfort no easy way to test this as has to be on master. But if you have a fix can merge.

```
/venv/local/lib/python2.7/site-packages/pandas/tests/io/test_stata.py:247: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/venv/local/lib/python2.7/site-packages/pandas/tests/io/test_stata.py:88: in read_dta
    return read_stata(file, convert_dates=True)
/venv/local/lib/python2.7/site-packages/pandas/io/stata.py:173: in read_stata
    data = reader.read()
/venv/local/lib/python2.7/site-packages/pandas/io/stata.py:1526: in read
    data = self._insert_strls(data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <pandas.io.stata.StataReader object at 0xede213ec>
data =       x    y            z
0   1.0  abc   4294967299
1   3.0  cba   8589934595
2  93.0       12884901891
    def _insert_strls(self, data):
        if not hasattr(self, 'GSO') or len(self.GSO) == 0:
            return data
        for i, typ in enumerate(self.typlist):
            if typ != 'Q':
                continue
>           data.iloc[:, i] = [self.GSO[k] for k in data.iloc[:, i]]
E           KeyError: 4294967299
/venv/local/lib/python2.7/site-packages/pandas/io/stata.py:1626: KeyError
```

jreback commented on pandas/io/stata.py:
```diff
@@ -816,9 +819,14 @@ def get_base_missing_value(cls, dtype):
+    def __init__(self, encoding='latin-1'):
+
+        if encoding not in VALID_ENCODINGS:
+            raise ValueError('Unknown encoding. Only latin-1 and  ascii '
```
extra space before ascii

jreback commented on pandas/tests/io/test_stata.py:
```diff
@@ -1276,3 +1276,9 @@ def test_out_of_range_float(self):
             tm.assertTrue('infinity' in cm.exception)
+
+    def test_invalid_encoding(self):
+        original = self.read_csv(self.csv3)
```
can you add the commit number as a comment

jreback reviewed (approved):
trivial comments. after its green you can push (and just ping me to merge)

bashtage committed:
BUG: Fix limited key range on 32-bit platofrms

Fix use of 64-bit integers as keys in general string objects (GSO) by
wrapping in strings when used as dictionary keys

bashtage commented:
@jreback I rolled a fix for the 32bit issue into this PR.  I can separate if you prefer.

jreback commented:
@bashtage totally fine here. thanks! ping on green.

jreback commented:
thanks! always a pleasure @bashtage

jreback commented:
@bashtage looks like this fixed the 32-bit error as well! https://travis-ci.org/MacPython/pandas-wheels/jobs/213622884

thanks (of course other errors which I am working on ......)
